### PR TITLE
Add a new site by providing just a shortname and (eventually) a production domain name

### PR DESCRIPTION
**Pull request by falkodev**
add shortname and production hostname
add task to convert old schema to new

### CODE REVIEW

File: lib/sites-base.js [hunk 1]
```diff
@@ -1,214 +1,308 @@
...
-          group = await apos.groups.insert(req, {
-            title: 'admin',
-            permissions: [ 'admin' ]
-          });
+    });
+  
+    self.addTask('transition-shortname', 'Convert first element of hostname array to shortname in order to use baseUrlDomains.', async function(apos, argv) {
+      const req = self.apos.tasks.getReq();
+      const sites = await self.find(req, {}).toArray();
+      for(const site of sites) {
```
Comment (boutell) [inline lib/sites-base.js]: Space before (

File: lib/sites-base.js [hunk 2]
```diff
@@ -1,214 +1,308 @@
...
+        if (!site.shortName && site.hostnamesArray && site.hostnamesArray.length) {
+          site.shortName = site.hostnamesArray[0].hostname;
```
Comment (boutell) [inline lib/sites-base.js]: This should be the *first part* of the hostname. That is, if the hostname is "foo.s", shortName should be "foo".

File: lib/sites-base.js [hunk 3]
```diff
@@ -1,214 +1,308 @@
...
-        await to.docs.db.remove({ trash: true });
-        await to.attachments.db.remove({ trash: true });
+      // Don't let an admin password wind up in the db in cleartext
+      req.newAdminPassword = doc.adminPassword;
+      delete doc.adminPassword;
+  
+      doc.hostnames = _.map(doc.hostnamesArray || [], function(value) {
+        return value.hostname.toLowerCase().trim();
+      });
+
```
Comment (boutell) [inline lib/sites-base.js]: Where is the logic to populate devBaseUrl, stagingBaseUrl, and prodBaseUrl from the baseUrlDomains property and shortName?

File: README.md [hunk 4]
```diff
@@ -289,7 +289,39 @@ VERBOSE=1 node app
 
 ## Setting `baseUrl` and naming environments
 
-By default, `sites` come with three url fields in their schema that correspond to three server environments: `dev`, `staging`, and `prod`. From the Dashboard, you are able to set the `baseUrl` property of each site within each environment. To let `apostrophe-multisite` know what environment it is currently running in, add the property `env` to your server's `data/local.js` file.
```
Comment (boutell) [inline README.md]: Actually the usual thing is to set the ENV environment variable to dev, staging or prod. local.js does work but multisite tends to be deployed to the cloud and clouds tend to prefer environment based configuration. So I'd document the ENV approach.

**Comment by boutell** [comment]
I recommend testing that things work as expected with ENV set to dev, staging and prod and appropriate /etc/hosts entries.

File: lib/sites-base.js (hunk 3 above)
Comment (falkodev) [reply lib/sites-base.js]: On line 95
`doc[key+'BaseUrl'] = shortname + '.' + self.options.baseUrlDomains[key].toLowerCase().trim()`

I started by writing 
`if (self.options.baseUrlDomains.dev) { ... } else if..` but I ended up iterating in the object `self.options.baseUrlDomains` and checking the key (that is to say 'dev', 'staging' or 'prod'). Would you prefer I add a comment or I explicitly add `if`s conditions?

File: lib/sites-base.js [hunk 5]
```diff
@@ -1,214 +1,326 @@
...
+      delete doc.adminPassword;
+  
+      if (self.options.baseUrlDomains) {
+        // new behavior
+        const shortname = doc.shortName || doc.slug;
+        doc.hostnames = []
 
-      let from, to;
+        _.forOwn(self.options.baseUrlDomains, function(value, key) {
+          doc[key+'BaseUrl'] = shortname + '.' + self.options.baseUrlDomains[key].toLowerCase().trim()
```
Comment (boutell) [inline lib/sites-base.js]: Spaces here:

key+'BaseUrl'

I'm saying there should be spaces around the +. The lack of spaces around the brackets is fine.

File: lib/sites-base.js [hunk 6]
```diff
@@ -1,214 +1,326 @@
...
+          return callback('adminPassword.required');
+        }
       }
+      // Don't let an admin password wind up in the db in cleartext
+      req.newAdminPassword = doc.adminPassword;
+      delete doc.adminPassword;
+  
+      if (self.options.baseUrlDomains) {
+        // new behavior
+        const shortname = doc.shortName || doc.slug;
```
Comment (boutell) [inline lib/sites-base.js]: Please be consistent in capitalization (shortName, not shortname).

File: lib/sites-base.js [hunk 7]
```diff
@@ -1,214 +1,326 @@
...
 
-      // copy the database and the attachments
-      await copyDatabase();
-      from = await getFrom();
-      to = await getTo();
-      await purgeTrash();
-      if (self.options.copy && self.options.copy.hardlink) {
-        await hardlinkAttachments();
+        if (doc.productionHostname) {
+          doc.hostnames.push(doc.productionHostname)
```
Comment (boutell) [inline lib/sites-base.js]: Statements should end with ;

File: README.md (hunk 4 above)
Comment (falkodev) [reply README.md]: Ok, I tried to follow your suggestion. If you have a better formulation, let me know please

**Comment by boutell** [comment]
That is a good question re iterating over the keys. I think we should make sure each key exists and apply them individually, but it is OK to check for just the existence of the parent key to turn on the overall feature. Some people might have just dev and prod.

**Pull request merged**